runtime/compiler/p/codegen/J9TreeEvaluator.cpp

```diff
@@ -1930,13 +1930,592 @@ TR::Register *J9::Power::TreeEvaluator::anewArrayEvaluator(TR::Node *node, TR::C
       return TR::TreeEvaluator::VMnewEvaluator(node, cg);
    }
 
+static TR::Register * generateMultianewArrayWithInlineAllocators(TR::Node *node, TR::CodeGenerator *cg)
+   {
+   TR::Compilation *comp = cg->comp();
+   TR_Debug *compDebug = comp->getDebug();
+   TR_ASSERT_FATAL(comp->target().is64Bit(), "multianewArrayEvaluator is only supported on 64-bit JVMs!");
+   TR_J9VMBase *fej9 = comp->fej9();
+   // size of reference field and address
+   int32_t referenceFieldSize = TR::Compiler->om.sizeofReferenceField();
+   TR_ASSERT_FATAL(referenceFieldSize <= 8,
+      "multianewArrayEvaluator - referenceFieldSize cannot be greater than 8!");
+   int32_t addrSize = TR::Compiler->om.sizeofReferenceAddress();
+   // the size of a length-0 array in bytes taking alignment into account
+   int32_t zeroArraySizeAligned = OMR::align(TR::Compiler->om.discontiguousArrayHeaderSizeInBytes(),
+                                             TR::Compiler->om.getObjectAlignmentInBytes());
+   // we use 64 bit classes only in 64 bit arch without compression
+   bool use64BitClasses = comp->target().is64Bit() && !TR::Compiler->om.generateCompressedObjectHeaders();
+   TR::InstOpCode::Mnemonic storeClassOp = use64BitClasses ? TR::InstOpCode::std : TR::InstOpCode::stw;
+   int8_t classPtrLen = use64BitClasses ? 8 : 4;
+
+   // Zero size arrays are considered "discontiguous", and the "mustBeZero" field
+   // of discontiguous arrays must be located where the "size" field of contiguous arrays is.
+   UDATA offsetOfMustBeZeroField = fej9->getOffsetOfContiguousArraySizeField();
+
+   // the maximum number of elements we can handle without risking an overflow
+   uintptr_t maxObjectSizeInElements = cg->getMaxObjectSizeGuaranteedNotToOverflow() / referenceFieldSize;
+
+   // When we allocate a block of pointers, we want to make sure all the block's size is aligned:
+   //    1. if the size of the referenceFields making up the block is aligned, do nothing;
+   //    2. otherwise, we first add refFieldSizeAligned-1 to the size, then mask the excess bits.
+   int32_t refFieldSizeAligned = OMR::align(referenceFieldSize,
+                                           TR::Compiler->om.getObjectAlignmentInBytes());
+   bool needsAlignRefField = (refFieldSizeAligned != referenceFieldSize);
+   int32_t alignmentCompensation = needsAlignRefField ? refFieldSizeAligned - 1 : 0;
+
+   // offHeap enabled
+#if defined(J9VM_GC_SPARSE_HEAP_ALLOCATION)
+   bool isOffHeapAllocationEnabled = TR::Compiler->om.isOffHeapAllocationEnabled();
+#endif /* defined(J9VM_GC_SPARSE_HEAP_ALLOCATION) */
+
+   // ptr to array of sizes, with the highest dimension in the front
+   TR::Register *dimsPtrReg = cg->evaluate(node->getFirstChild());
+   // number of dimensions - compile time constant
+   uint32_t nDims = node->getSecondChild()->get32bitIntegralValue();
+   // class pointer of objects in the array - in the 2D case this is the class of the 1D array
```

Comment:
reference to 1st D or 2nd D is ambiguous. maybe top level vs. bottom level is clearer? or, clarify it with 1st D (or 1D) referring to the top level?